ksql-engine/src/main/java/io/confluent/ksql/topic/DefaultTopicInjector.java

```diff
@@ -0,0 +1,153 @@
+/*
+ * Copyright 2019 Confluent Inc.
+ *
+ * Licensed under the Confluent Community License; you may not use this file
+ * except in compliance with the License.  You may obtain a copy of the License at
+ *
+ * http://www.confluent.io/confluent-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.topic;
+
+import io.confluent.ksql.KsqlExecutionContext;
+import io.confluent.ksql.analyzer.Analysis;
+import io.confluent.ksql.analyzer.QueryAnalyzer;
+import io.confluent.ksql.ddl.DdlConfig;
+import io.confluent.ksql.metastore.MetaStore;
+import io.confluent.ksql.metastore.StructuredDataSource;
+import io.confluent.ksql.parser.KsqlParser.PreparedStatement;
+import io.confluent.ksql.parser.SqlFormatter;
+import io.confluent.ksql.parser.tree.CreateAsSelect;
+import io.confluent.ksql.parser.tree.Expression;
+import io.confluent.ksql.parser.tree.IntegerLiteral;
+import io.confluent.ksql.parser.tree.Statement;
+import io.confluent.ksql.parser.tree.StringLiteral;
+import io.confluent.ksql.services.KafkaTopicClient;
+import io.confluent.ksql.util.KsqlConfig;
+import io.confluent.ksql.util.KsqlConstants;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import org.apache.commons.lang3.ObjectUtils;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.kafka.clients.admin.TopicDescription;
+
+public class DefaultTopicInjector implements TopicInjector {
+
+  private final KafkaTopicClient topicClient;
+  private final MetaStore metaStore;
+
+  public DefaultTopicInjector(
+      final KsqlExecutionContext executionContext
+  ) {
+    this(executionContext.getServiceContext().getTopicClient(), executionContext.getMetaStore());
+  }
+
+  DefaultTopicInjector(
+      final KafkaTopicClient topicClient,
+      final MetaStore metaStore) {
+    this.topicClient = topicClient;
+    this.metaStore = metaStore;
+  }
+
+  @SuppressWarnings("unchecked")
+  @Override
+  public <T extends Statement> PreparedStatement<T> forStatement(
+      final PreparedStatement<T> statement,
+      final KsqlConfig ksqlConfig,
+      final Map<String, Object> propertyOverrides) {
+    if (!(statement.getStatement() instanceof CreateAsSelect)) {
+      return statement;
+    }
+
+    final CreateAsSelect cas = (CreateAsSelect) statement.getStatement();
+
+    final Analysis analysis = new QueryAnalyzer(
+        metaStore,
+        ksqlConfig.getString(KsqlConfig.KSQL_OUTPUT_TOPIC_NAME_PREFIX_CONFIG))
+        .analyze(SqlFormatter.formatSql(cas), cas.getQuery(), Optional.of(cas.getSink()));
```

Comment:
I started down this path, then got very nervous. There's a non-negligible amount of not so straightforward logic that would need to be replicated if I wanted to build my own (or adapt `PlanSourceExtractorVisitor`) and my confidence that it would work with complicated nested queries and joins while maintaining "get me true source" logic is small. 

Plus, if anyone were ever to make a change to the QueryAnalyzer, they would need to know to change the new extractor as well (assuming unit tests didn't catch whatever change they made). 

When push comes to shove, this code _does_ analyze the entire query structure, even if it is just to extract a small part of it, so I don't believe it is wrong. Performance shouldn't be a concern given where it is, and while this does feel like overkill I would rather use a robust component then rely on an ad-hoc bootstrapped on that I build and then need to maintain.